File: document/manager.go
// Package document provides document state management.
package document

import "sync"

// Document represents an open text document.
type Document struct {
	URI        string
	LanguageID string
	Version    int
	Content    string
}

// Manager manages open documents.
type Manager struct {
	mu        sync.RWMutex
	documents map[string]*Document
}

// NewManager creates a new document manager.
func NewManager() *Manager {
	return &Manager{
		documents: make(map[string]*Document),
	}
}

// Open opens a document and stores it in the manager.
func (m *Manager) Open(uri, languageID string, version int, content string) {
	m.mu.Lock()
	defer m.mu.Unlock()

	m.documents[uri] = &Document{
		URI:        uri,
		LanguageID: languageID,
		Version:    version,
		Content:    content,
	}
}

// Change updates the content of an open document.
// For full sync mode, the entire content is replaced.
func (m *Manager) Change(uri string, version int, content string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()

	doc, ok := m.documents[uri]
	if !ok {
		return false
	}

	doc.Version = version
	doc.Content = content
	return true
}

// Close removes a document from the manager.
func (m *Manager) Close(uri string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()

	if _, ok := m.documents[uri]; !ok {
		return false
	}

	delete(m.documents, uri)
	return true
}

// Get retrieves a document by URI.
func (m *Manager) Get(uri string) (*Document, bool) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	doc, ok := m.documents[uri]
	if !ok {
		return nil, false
	}

	// Return a copy to avoid race conditions
	return &Document{
		URI:        doc.URI,
		LanguageID: doc.LanguageID,
		Version:    doc.Version,
		Content:    doc.Content,
	}, true
}

// All returns all open documents.
func (m *Manager) All() []*Document {
	m.mu.RLock()
	defer m.mu.RUnlock()

	docs := make([]*Document, 0, len(m.documents))
	for _, doc := range m.documents {
		docs = append(docs, &Document{
			URI:        doc.URI,
			LanguageID: doc.LanguageID,
			Version:    doc.Version,
			Content:    doc.Content,
		})
	}
	return docs
}

File: lsp_server/document/manager.py
"""Document manager for tracking open text documents."""
from __future__ import annotations

import copy
import threading
from dataclasses import dataclass
from typing import Optional


@dataclass
class Document:
    uri: str
    language_id: str
    version: int
    content: str


class Manager:
    """Thread-safe manager for open text documents."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._documents: dict[str, Document] = {}

    def open(self, uri: str, language_id: str, version: int, content: str) -> None:
        """Store a newly opened document."""
        with self._lock:
            self._documents[uri] = Document(
                uri=uri,
                language_id=language_id,
                version=version,
                content=content,
            )

    def change(self, uri: str, version: int, content: str) -> bool:
        """Update an existing document. Returns True if the document existed."""
        with self._lock:
            if uri not in self._documents:
                return False
            doc = self._documents[uri]
            doc.version = version
            doc.content = content
            return True

    def close(self, uri: str) -> bool:
        """Remove a document. Returns True if the document existed."""
        with self._lock:
            if uri not in self._documents:
                return False
            del self._documents[uri]
            return True

    def get(self, uri: str) -> Optional[Document]:
        """Get a copy of a document by URI, or None if not found."""
        with self._lock:
            doc = self._documents.get(uri)
            if doc is None:
                return None
            return copy.copy(doc)

    def all(self) -> list[Document]:
        """Get copies of all open documents."""
        with self._lock:
            return [copy.copy(d) for d in self._documents.values()]
